Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm

```diff
@@ -59,11 +60,226 @@
+        return NSAccessibilityGroupRole;
+    if ([attribute isEqualToString:NSAccessibilityPrimaryScreenHeightAttribute])
+        return [[self parent] accessibilityAttributeValue:NSAccessibilityPrimaryScreenHeightAttribute];
+    if ([attribute isEqualToString:NSAccessibilitySubroleAttribute])
+        return @"AXPDFPluginSubrole";
+    if ([attribute isEqualToString:NSAccessibilitySizeAttribute])
+    {
+        RefPtr plugin = _pdfPlugin.get();
+        if (plugin)
+            return [NSValue valueWithSize:plugin->boundsOnScreen().size()];
+    }
+    if ([attribute isEqualToString:NSAccessibilityPositionAttribute])
+    {
+        RefPtr plugin = _pdfPlugin.get();
+        if (plugin)
+            return [NSValue valueWithPoint:plugin->boundsOnScreen().location()];
+    }
+    return nil;
+}
+
+ALLOW_DEPRECATED_IMPLEMENTATIONS_BEGIN
+- (NSArray *)accessibilityAttributeNames
+ALLOW_DEPRECATED_IMPLEMENTATIONS_END
+{
+    static NeverDestroyed<RetainPtr<NSArray>> attributeNames = @[
+        NSAccessibilityParentAttribute,
+        NSAccessibilityWindowAttribute,
+        NSAccessibilityTopLevelUIElementAttribute,
+        NSAccessibilityRoleDescriptionAttribute,
+        NSAccessibilitySizeAttribute,
+        NSAccessibilityEnabledAttribute,
+        NSAccessibilityPositionAttribute,
+        NSAccessibilityFocusedAttribute,
+        NSAccessibilityChildrenAttribute,
+        NSAccessibilityPrimaryScreenHeightAttribute,
+        NSAccessibilitySubroleAttribute
+    ];
+    return attributeNames.get().get();
+}
+
+- (BOOL)accessibilityShouldUseUniqueId
+{
+    return YES;
+}
+
+- (NSUInteger)accessibilityArrayAttributeCount:(NSString *)attribute
+{
+    if (!_pdfDocument) {
+        RefPtr plugin = _pdfPlugin.get();
+        if (plugin)
+            _pdfDocument = plugin->pdfDocument();
+    }
+    if ([attribute isEqualToString:NSAccessibilityChildrenAttribute])
+        return [_pdfDocument.get() pageCount];
+    if ([attribute isEqualToString:NSAccessibilityVisibleChildrenAttribute])
+        return [self accessibilityVisibleChildren].count;
+    return [super accessibilityArrayAttributeCount:attribute];
+}
+
+- (NSArray*)accessibilityChildren
+{
+    if (!_pdfDocument) {
+        RefPtr plugin = _pdfPlugin.get();
+        if (plugin)
+            _pdfDocument = plugin->pdfDocument();
+    }
+    return [_pdfDocument.get() accessibilityChildren:self];
+}
+
+- (id)accessibilityHitTest:(NSPoint)point
+{
+    for (id element in [self accessibilityChildren]) {
+        id result = [element accessibilityHitTest:point];
+        if (!result)
+            return result;
+    }
+    return self;
+}
+
+// this function allows VoiceOver to scroll to the current page with VO cursor
+- (void)gotoDestination:(PDFDestination *)destination
+{
+    NSInteger pageIndex = [_pdfDocument indexForPage:[destination page]];
+    PDFRect rect = [[self accessibilityChildren][0] accessibilityFrame];
```

Comment:
I think we should compute the destination scroll position in the main thread code; this computation is almost certainly wrong with rotated pages, zooming etc. We already have `UnifiedPDFPlugin::scrollToPage()` that you can call.

Reply:
added a Fixme and will fix that once I have the separate file for accessibility object